pkgs/applications/editors/eclipse/plugins.nix

```diff
@@ -15,80 +34,123 @@ rec {
 
       passthru = {
         isEclipsePlugin = true;
+        dependencies = [];
       } // passthru;
     });
 
-  # Helper for the common case where we have separate feature and
-  # plugin JARs.
-  buildEclipsePlugin =
-    { name, srcFeature, srcPlugin ? null, srcPlugins ? [], ... } @ attrs:
-      assert srcPlugin == null -> srcPlugins != [];
-      assert srcPlugin != null -> srcPlugins == [];
-
-      let
-
-        pSrcs = if (srcPlugin != null) then [ srcPlugin ] else srcPlugins;
-
-      in
-
-        buildEclipsePluginBase (attrs // {
-          srcs = [ srcFeature ] ++ pSrcs;
-
-          buildCommand = ''
-            dropinDir="$out/eclipse/dropins/${name}"
-
-            mkdir -p $dropinDir/features
-            unzip ${srcFeature} -d $dropinDir/features/
+  # Mirror a subset of a p2 update site containing the given features and
+  # plugins.
+  mirrorEclipsePlugins = { pname
+                         , version
+                         , repository ? null
+                         , artifactRepository ? null
+                         , metadataRepository ? null
+                         , installableUnit ? null
+                         , installableUnits ? null
+                         , eclipse ? eclipses.eclipse-sdk
+                         , hash
+                         , ... } @ attrs:
+    assert (artifactRepository == null) == (metadataRepository == null);
+    assert (repository == null) != (artifactRepository == null);
+    assert (installableUnit == null) != (installableUnits == null);
+
+    let
+      ius =
+        if installableUnit != null
+        then [ installableUnit ]
+        else installableUnits;
+      iuEntries = (lib.strings.concatStrings (map (iu: "<iu id=\"${iu.id}\" version=\"${iu.version}\" />") ius));
+
+    in
+      buildEclipsePluginBase rec {
+        name = "${pname}-${version}";
+        inherit version;
+
+        buildCommand = ''
```

Comment:
There is a simpler way to mirror repositories than using an ant task... but unfortunately it mirrors the entire repository which means we can't have a reliable output hash. The ant task allows us to "slice" the repository, i.e. include only the specified installable units (features/plugins).